### PR TITLE
Only suggest turbofish in patterns if we may recover

### LINKED ISSUE #115780: regression: generic args in patterns require the turbofish syntax
```
[INFO] [stdout] error: generic args in patterns require the turbofish syntax
[INFO] [stdout]    --> src/lib.rs:142:16
[INFO] [stdout]     |
[INFO] [stdout] 142 |             ret<TCon> (Clone::clone(x_val), Clone::clone(y_val));
[INFO] [stdout]     |                ^
[INFO] [stdout]     |
[INFO] [stdout]    ::: src/mdo.rs:19:9
[INFO] [stdout]     |
[INFO] [stdout] 19  |         $p: pat =<< $e: expr ; $( $t: tt )*
[INFO] [stdout]     |         ------- while parsing argument for this `pat` macro fragment
[INFO] [stdout]     |
[INFO] [stdout] help: use `::<...>` instead of `<...>` to specify lifetime, type, or const arguments
[INFO] [stdout]     |
[INFO] [stdout] 142 |             ret::<TCon> (Clone::clone(x_val), Clone::clone(y_val));
[INFO] [stdout]     |                ++
[INFO] [stdout] 
[INFO] [stdout] 
```

https://crater-reports.s3.amazonaws.com/beta-1.73-1.2/beta-2023-09-10/reg/haskell_bits-0.3.0/log.txt

### PR DESCRIPTION
**Pull request by fmease**
Fixes [after backport] #115780.

CC #103534.

### CODE REVIEW

File: tests/ui/parser/issues/issue-115780-pat-lt-bracket-in-macro-call.rs
Comment (fmease) [inline tests/ui/parser/issues/issue-115780-pat-lt-bracket-in-macro-call.rs:1]: Originally I wanted to place this file into `parser/` directly but its file entry limit of 900 has been reached

Comment (fmease) [reply tests/ui/parser/issues/issue-115780-pat-lt-bracket-in-macro-call.rs:1]: Naming of this file mirrors naming in #114300.

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks for fixing the regression so quickly @fmease!

@bors r+

**Comment by compiler-errors** [comment]
This should definitely be beta backported, imo

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-09-14/near/390942105)

@rustbot label +beta-accepted

**Pull request merged**